nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/GetSolr.java

```diff
@@ -170,159 +210,213 @@ protected void init(final ProcessorInitializationContext context) {
 
-                    while (endRow < totalResults) {
-                        solrQuery.setStart(endRow);
+            solrQuery.setParam(CursorMarkParams.CURSOR_MARK_PARAM, context.getStateManager().getState(Scope.CLUSTER).get(STATE_MANAGER_CURSOR_MARK));
+            solrQuery.setRows(context.getProperty(BATCH_SIZE).asInteger());
 
-                        stopWatch.start();
-                        response = getSolrClient().query(solrQuery);
-                        stopWatch.stop();
+            final StringBuilder sortClause = (new StringBuilder())
+                    .append(dateField)
+                    .append(" asc,")
+                    .append(id_field)
+                    .append(" asc");
+            solrQuery.setParam("sort", sortClause.toString());
 
-                        duration = stopWatch.getDuration(TimeUnit.MILLISECONDS);
-                        logger.info("Retrieved results for {} in {} ms", new Object[]{query, duration});
+            while (continuePaging.get()) {
+                final QueryRequest req = new QueryRequest(solrQuery);
+                if (isBasicAuthEnabled()) {
+                    req.setBasicAuthCredentials(getUsername(), getPassword());
+                }
 
-                        flowFile = session.create();
+                logger.debug(solrQuery.toQueryString());
+                final QueryResponse response = req.process(getSolrClient());
+                final SolrDocumentList documentList = response.getResults();
+
+                if (response.getResults().size() > 0) {
+                    final SolrDocument lastSolrDocument = documentList.get(response.getResults().size()-1);
+                    final String latestDateValue = df.format(lastSolrDocument.get(dateField));
+
+                    solrQuery.setParam(CursorMarkParams.CURSOR_MARK_PARAM, response.getNextCursorMark());
+                    final Map<String,String> updateStateManager = new HashMap<String,String>();
+                    updateStateManager.putAll(context.getStateManager().getState(Scope.CLUSTER).toMap());
+                    updateStateManager.put(STATE_MANAGER_CURSOR_MARK, response.getNextCursorMark());
+                    updateStateManager.put(STATE_MANAGER_FILTER, latestDateValue);
+                    context.getStateManager().setState(updateStateManager, Scope.CLUSTER);
+
+                    FlowFile flowFile = session.create();
+                    flowFile = session.putAttribute(flowFile, "solrQuery", solrQuery.toString());
+
+                    if (context.getProperty(RETURN_TYPE).getValue().equals(MODE_XML.getValue())){
+                        if (dateFieldNotInSpecifiedFieldsList.get()) {
+                            for (SolrDocument doc : response.getResults()) {
+                                doc.removeFields(dateField);
+                            }
+                        }
                         flowFile = session.write(flowFile, new QueryResponseOutputStreamCallback(response));
```

Comment:
Surprisingly, this processor has not been able to output a proper XML if multiple documents are fetched at a single onTrigger method. The resulted FlowFile contains following text data, but it's not a well formatted XML and can't be parsed at downstream correctly. We should wrap `doc` elements by something like `docs` to make it parsable..

```
<doc boost="1.0">
<field name="id">F8V7067-APL-KIT</field>
<field name="price_c____l_ns">1995</field>
</doc>
<doc boost="1.0">
<field name="id">MA147LL/A</field>
</doc>
```

Would you be able to fix this method, too?
https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/GetSolr.java#L339

Reply:
Sure. But all users will have to change their workflows after updating NiFi, right?

Reply:
Yes, that's correct, but it's necessary.